mat/svd_test.go

```diff
@@ -212,3 +212,451 @@ func extractSVD(svd *SVD) (s []float64, u, v *Dense) {
+		a    [][]float64
+		b    []float64
+		cond float64
+		ans  []float64
+	}{
+		{
+			a:   [][]float64{{6}},
+			b:   []float64{3},
+			ans: []float64{0.5},
+		},
+		{
+			a: [][]float64{
+				{1, 0, 0},
+				{0, 1, 0},
+				{0, 0, 1},
+			},
+			b:   []float64{3, 2, 1},
+			ans: []float64{3, 2, 1},
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.5528},
+				{0.9058, 0.6324, 0.8723},
+				{0.1270, 0.0975, 0.7612},
+			},
+			b:   []float64{0.278, 0.547, 0.958},
+			ans: []float64{-0.932687281002860, 0.303963920182067, 1.375216503507109},
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.5528},
+				{0.9058, 0.6324, 0.8723},
+			},
+			b:   []float64{0.278, 0.547},
+			ans: []float64{0.25919787248965376, -0.25560256266441034, 0.5432324059702451},
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.9},
+				{0.9058, 0.6324, 0.9},
+				{0.1270, 0.0975, 0.1},
+				{1.6, 2.8, -3.5},
+			},
+			b:   []float64{0.278, 0.547, -0.958, 1.452},
+			ans: []float64{0.820970340787782, -0.218604626527306, -0.212938815234215},
+		},
+		{
+			// Test rank-deficient case compared with numpy.
+			// >>> import numpy as np
+			// >>> b = np.array([[-2.3181340317357653],
+			// ...     [-0.7146777651358073],
+			// ...     [1.8361340927945298],
+			// ...     [-0.35699930593018775],
+			// ...     [-1.6359508076249094]])
+			// >>> A = np.array([[-1.7854591879711257, -0.42687285925779594, -0.12730256811265162],
+			// ...     [-0.5728984211439724, -0.10093393134001777, -0.1181901192353067],
+			// ...     [1.2484316018707418, 0.5646683943038734, -0.48229492403243485],
+			// ...     [0.10174927665169475, -0.5805410929482445, 1.3054473231942054],
+			// ...     [-1.134174808195733, -0.4732430202414438, 0.3528489486370508]])
+			// >>> np.linalg.lstsq(A, b, rcond=None)
+			// (array([[ 1.21208422],
+			//        [ 0.41541503],
+			//        [-0.18320349]]), array([], dtype=float64), 2, array([2.68451480e+00, 1.52593185e+00, 6.82840229e-17]))
+
+			a: [][]float64{
+				{-1.7854591879711257, -0.42687285925779594, -0.12730256811265162},
+				{-0.5728984211439724, -0.10093393134001777, -0.1181901192353067},
+				{1.2484316018707418, 0.5646683943038734, -0.48229492403243485},
+				{0.10174927665169475, -0.5805410929482445, 1.3054473231942054},
+				{-1.134174808195733, -0.4732430202414438, 0.3528489486370508},
+			},
+			b:    []float64{-2.3181340317357653, -0.7146777651358073, 1.8361340927945298, -0.35699930593018775, -1.6359508076249094},
+			cond: 1e-15,
+			ans:  []float64{1.2120842180372118, 0.4154150318658529, -0.1832034870198265},
+		},
+		{
+			a: [][]float64{
+				{0, 0},
+				{0, 0},
+			},
+			b: []float64{3, 2},
+		},
+		{
+			a: [][]float64{
+				{0, 0},
+				{0, 0},
+				{0, 0},
+			},
+			b: []float64{3, 2, 1},
+		},
+		{
+			a: [][]float64{
+				{0, 0, 0},
+				{0, 0, 0},
+			},
+			b: []float64{3, 2},
+		},
+	} {
+		a := NewDense(flatten(test.a))
+		b := NewVecDense(len(test.b), test.b)
+
+		var ans *VecDense
+		if test.ans != nil {
```

Comment:
How does this work? What is being tested if `ans` is nil?

Reply:
Nothing. The SVD factorisation has failed.